onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py

```diff
@@ -0,0 +1,463 @@
+
+            bbox = (detection.xmin, detection.ymin,
+                    detection.xmax, detection.ymax)
+
+            label_idx = detection.label
+            label = self.classes[label_idx]
+
+            confidence = detection.confidence
+            # x is left/right axis, where 0 is in middle of the frame,
+            #   to the left is negative x, and to the right is positive x
+            # y is down/up axis, where 0 is in the middle of the frame,
+            # down is negative y, and up is positive y z is distance of
+            #   object from camera in mm
+            x_cam_mm = detection.spatialCoordinates.x
+            y_cam_mm = detection.spatialCoordinates.y
+            z_cam_mm = detection.spatialCoordinates.z
+
+            # Get sonar sweep range
+            (left_end, right_end) = coords_to_angle(
+                (detection.xmin - 0.5) * CAMERA_PIXEL_WIDTH,
+                (detection.xmax - 0.5) * CAMERA_PIXEL_WIDTH)
+
+            left_end = float(left_end)
+            right_end = float(right_end)
+
+            # Rescale left and right end points to fit better sonar image
+            left_end = left_end * 0.5 + 5
+            right_end = right_end * 0.5 + 5
+
+            x_cam_meters = mm_to_meters(x_cam_mm)
+            y_cam_meters = mm_to_meters(y_cam_mm)
+            z_cam_meters = mm_to_meters(z_cam_mm)
+
+            det_coords_robot_mm = camera_frame_to_robot_frame(
+                    x_cam_meters, y_cam_meters, z_cam_meters)
+
+            # Create a new sonar request msg object
+            sonar_request_msg = sweepGoal()
+            sonar_request_msg.start_angle = left_end
+            sonar_request_msg.end_angle = right_end
+            sonar_request_msg.distance_of_scan = SONAR_DEPTH
+            self.sonar_requests_publisher.publish(sonar_request_msg)
+
+            # Try calling sonar on detected bounding box
+            # if sonar responds, then override existing robot-frame x, y info;
+            # else, keep default
+            if not (self.sonar_response == (0, 0)) and self.in_sonar_range:
+                det_coords_robot_mm = (self.sonar_response[0],
+                                       self.sonar_response[1], y_cam_meters)
+
+                # Publish Pose to task planning directly; based on
+                # latest sonar update
+                published_pose = Pose()
+                published_pose.position.x = det_coords_robot_mm[0]
+                published_pose.position.y = det_coords_robot_mm[1]
+                published_pose.position.z = det_coords_robot_mm[2]
+                published_pose.orientation.x = 0
+                published_pose.orientation.y = 0
+                published_pose.orientation.z = 0
+                published_pose.orientation.w = 1
+
+            self.pose_publisher.publish()
+            self.publish_prediction(
+                bbox, det_coords_robot_mm, label, confidence,
+                (height, width), self.in_sonar_range)
```

Comment:
Even if sonar gives an error by returning (0, 0), this might still think it's using sonar

Reply:
Currently in_sonar_range will only ever be true if sonar is running and in range, so the only case for if sonar_response is (0, 0) is when the sonar has not responded, which in case in_sonar_range will be False.